File: golang/sort_array_by_parity.go
package main

// Space   : O(1)
// Time    : O(n)

func sortArrayByParity(A []int) []int {
	leng := len(A)
	idx1, idx2 := 0, leng-1
	ans := make([]int, leng)

	for i := 0; i < leng; i++ {
		if A[i]%2 == 0 {
			ans[idx1] = A[i]
			idx1++
		} else {
			ans[idx2] = A[i]
			idx2--
		}
	}

	return ans
}

File: python3/sort_array_by_parity.py
"""
Space   : O(n)
Time    : O(n log n)
"""


class Solution:
    def sortArrayByParity(self, A: List[int]) -> List[int]:
        ev, od = [], []

        for x in A:
            if x % 2 == 0:
                ev.append(x)
            else:
                od.append(x)

        return sorted(ev) + sorted(od)
